extensions/big-data-cluster/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,376 @@
+		let controllerNode = this.getExistingControllerNode(url, username);
+		if (controllerNode) {
+			controllerNode.password = password;
+			controllerNode.rememberPassword = rememberPassword;
+			controllerNode.clearChildren();
+		} else {
+			controllerNode = new ControllerNode({ url, username, password, rememberPassword, parent: this, treeChangeHandler: this.treeChangeHandler });
+			this.addChild(controllerNode);
+		}
+
+		if (masterInstance) {
+			controllerNode.addSqlMasterNode(masterInstance.endpoint, masterInstance.description);
+		}
+	}
+
+	public deleteControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		let index = nodes.findIndex(e => e.url === url && e.username === username);
+		let deleted = undefined;
+		if (index >= 0) {
+			deleted = nodes.splice(index, 1);
+		}
+		return deleted;
+	}
+
+	private getExistingControllerNode(url: string, username: string): ControllerNode {
+		if (!url || !username) {
+			return undefined;
+		}
+		let nodes = this.children as ControllerNode[];
+		return nodes.find(e => e.url === url && e.username === username);
+	}
+}
+
+export type ControllerNodeArg = {
+	url: string,
+	username: string,
+	password: string,
+	parent: ControllerTreeNode,
+	treeChangeHandler: IControllerTreeChangeHandler,
+	label?: string,
+	description?: string,
+	rememberPassword?: boolean
+};
+
+export class ControllerNode extends ControllerTreeNode {
+	private _url: string;
+	private _username: string;
+	private _password: string;
+	private _rememberPassword: boolean;
+
+	constructor(arg?: ControllerNodeArg) {
+		super(Object.assign({
+			label: undefined,
+			nodeType: BdcItemType.controller,
+			iconPath: IconPath.controllerNode
+		}, arg));
+
+		let address = ControllerNode.toIpAndPort(arg.url);
+		this.label = arg.label || `controller: ${address} (${arg.username})`;
+		this.description = arg.description || this.label;
+		this._url = arg.url;
+		this._username = arg.username;
+		this._password = arg.password;
+		this._rememberPassword = !!arg.rememberPassword;
+	}
+
+
+	public async getChildren(): Promise<ControllerTreeNode[]> {
+		if (this.children && this.children.length > 0) {
+			this.clearChildren();
+		}
+
+		if (!this._password) {
+			vscode.commands.executeCommand('bigDataClusters.command.addController', this);
+			return this.children as ControllerTreeNode[];
+		}
+
+		try {
+			let clusterController = new ClusterController();
+			let response = await clusterController.getEndPoints(this._url, this._username, this._password, true);
+			if (response && response.endPoints) {
+				let master = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+				this.addSqlMasterNode(master.endpoint, master.description);
+			}
+			return this.children as ControllerTreeNode[];
+		} catch (error) {
+			let e = error as IControllerError;
+			vscode.window.showErrorMessage(e.message);
```

Comment:
In other cases we also show the error code. Shouldn't we be consistent? Also what if it's a string error?

Reply:
@kevcunnane This is actually something I've been thinking about lately. It's best practice to never throw a non-Error object and I think we've done a pretty good job of that and so generally I've just been fine with people assuming it's an Error object and leaving it at that. Thoughts on doing that? I hate having code that doesn't really do anything valuable (checking if it's an error or not) 99.99% of the time, but I suppose if that 0.01% of the time causes broken behavior than maybe it's worth it just to always check?

Reply:
Yeah my concern is I've seen it happen much too often. Especially if calling into external modules. Hence I'd rather be safe than "right". I'm open either way, if we're confident it can't be a string then keep it simple.

Reply:
fixed.